design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,309 @@
+
+    <!-- Product Service -->
+    <service class="Magento\Catalog\Api\ProductRepositoryInterface" method="save"/>
+</services>
+```
+ 
+We will have to pass authenticated user information to remote services so we'll add _X-Magento-User-Type_ and _X-Magento-User-Id_
+headers to requests for a _UserContextInterface_ instance to be initialized. There's no need for nodes to validate auth information of
+incoming
+requests since those requests will come from other nodes in a closed network or they will be encrypted. These headers will only be
+accepted by endpoints declared in _inter_webapi.xml_ files.
+ 
+The way exceptions thrown by services are serialized right now for RESTful web API needs to be changed for inter-service endpoints.
+Exceptions are serialized without their type information provided - this info is crutial for services to communicate with each other so
+the type information will be added to the output.
+ 
+_inter_webapi.xml_ files will be contained in separate <ModuleName>InterApi modules and will have to be installed
+on nodes accepting remote inter-service requests.
+ 
+##### Changes for retries mechanism
+The invoker implementation based on RESTful web APIs will offer default retries mechanism, but developers will be able to disable it
+by using _di.xml_ (see Invoker API and SPI [here](invoker.md)) for when they would want to delegate this responsibility
+to a service mash or a balancer.
+ 
+When default retries mechanism is enabled _X-Magento-Call-Id_ header will be added to inter-service requests to identify calls.
+It will be randomly generated. Then if a timeout reached while waiting for a service to respond the invoker will send the same request
+again configured number of times with the same call ID. The node receiving requests will store the call ID in cache as a key and
+assign an empty value to it. When the node will have the service call result it will store it with the call ID. That way when a
+duplicate request comes in if the node can't find the call ID in cache it will execute the requested service method as usual,
+if the cache has the call ID that would mean the service method is being executed and we'll have to wait until the original process
+finishes it and writes the response to cache, if both call ID and the result are in cache then we just return the result. That way
+we will avoid executing duplicate calls when retries mechanism is applied.
+ 
+The timeout and tries limit will be configured via di.xml (see the invoker API and SPI [here](invoker.md)).
+    
+#### Service mashes and balancers
+Since we going to contact services via HTTP it is possible for configured service base URLs to actually lead to service mashes or
+balancers - that way the retries and lookup mechanism may be delegated.
+
+#### MQ RPC for gateway
+Magento has another way to remotely call services - RPC via a message queue. The problem with it is that right now it can only work
```

Comment:
Could you elaborate more on this. What kind of gateway you propose to use message queue for?